python/cusignal/io/reader.py

```diff
@@ -0,0 +1,328 @@
+
+    if num_samples < 0 or offset < 0:
+        raise ValueError("'num_samples' and 'offset' must be >= 0")
+
+    if dtype is None and (num_samples != 0 or offset != 0):
+        raise TypeError(
+            "'dtype' must be provided is 'num_samples'/'offset' != 0"
+        )
+
+    num_bytes_read = num_samples * cp.dtype(dtype).itemsize
+    offset *= cp.dtype(dtype).itemsize
+
+    if (offset % resource.getpagesize()) != 0:
+        raise ValueError("'offset' must be a multiple of the PAGESIZE")
+
+    with open(file, "rb") as f:
+        mm = mmap(
+            f.fileno(),
+            num_bytes_read,
+            flags=MAP_PRIVATE,
+            prot=PROT_READ,
+            offset=offset,
+        )
+        out = cp.asarray(mm)
+        stream.synchronize()
+        mm.close()
+
+    return out
+
+
+def write_bin(file, binary, buffer=None, append=True):
+    """
+    Writes binary array to file.
+
+    Parameters
+    ----------
+    file : str
+        A string of filename to store output.
+    binary : ndarray
+        Binary array to be written to file.
+    buffer : ndarray, optional
+        Pinned memory buffer to use when copying data from GPU.
+    append : bool, optional
+        Append to file if created.
+
+    Returns
+    -------
+    out : ndarray
+        An 1-dimensional array containing binary data.
+
+    """
+
+    # Get current stream, default or not.
+    stream = cp.cuda.get_current_stream()
+
+    if buffer is None:
+        buffer = cp.asnumpy(binary)
+    else:
+        binary.get(out=buffer)
+
+    if append is True:
+        mode = "ab"
+    else:
+        mode = "wb"
+
+    with open(file, mode) as f:
+        stream.synchronize()
+        f.write(buffer)
+
+
+def unpack_bin(binary, dtype, endianness="L"):
+    """
+    Unpack binary file. If endianness is big-endian, it my be converted
+    to little endian for NVIDIA GPU compatibility.
+
+    Parameters
+    ----------
+    binary : ndarray
+        The binary array to be unpack.
+    dtype : data-type, optional
+        Any object that can be interpreted as a numpy data type.
+    endianness : {'L', 'B'}, optional
+        Data set byte order
+
+    Returns
+    -------
+    out : ndarray
+        An 1-dimensional array containing unpacked binary data.
+
+    """
+
+    if endianness != "L" and endianness != "B" and endianness != "N":
+        raise ValueError("'endianness' should be 'L' or 'B'")
+
+    out = _unpack(binary, dtype, endianness)
+
+    return out
+
+
+def pack_bin(in1):
+    """
+    Pack binary arrary.
+    Data will be packed with little endian for NVIDIA GPU compatibility.
+
+    Parameters
+    ----------
+    in1 : ndarray
+        The ndarray to be pack at binary.
+
+    Returns
+    -------
+    out : ndarray
+        An 1-dimensional array containing packed binary data.
+
+    """
+
+    out = _pack(in1)
+
+    return out
+
+
+def read_sigmf(file, num_samples=0, offset=0):
+    """
+    Read and unpack binary file to GPU memory
```

Comment:
Update to SigMF

Reply:
???

Reply:
Whoops, your description has “read and unpack binary file” rather than “read and unpack sigmf file”

Reply:
Ohhh. Gotcha.